pkgs/applications/networking/bee/bee-clef.nix

```diff
@@ -0,0 +1,73 @@
+{ version ? "release", stdenv, fetchFromGitHub, go-ethereum, pkgs }:
+
+let
+
+  versionSpec = {
+    release = rec {
+      name = "bee-clef";
+      version = "0.4.4";
+      rev = "refs/tags/v${version}";
+      sha256 = "1bmgbav52pcb5p7cgq9756512fzfqhjybyr0dv538plkqx47mpv7";
+    };
+  }.${version};
+
+in
+
+stdenv.mkDerivation {
+  pname = versionSpec.name;
+  version = versionSpec.version;
+
+  # TODO delme
+  src = fetchFromGitHub {
+      owner = "attila-lendvai";
+      repo = "bee-clef";
+      rev = "710768de95e0521e6d32aceaf2b27d718a6ffea3";
+      sha256 = "1kn12zdsf341mrcdwnvx159v2pfarl13xz9izl06lhhbsa8gxnjx";
+  };
+
+  # src = fetchFromGitHub {
+  #   owner = "ethersphere";
+  #   repo = "bee-clef";
+  #   rev = versionSpec.rev;
+  #   sha256 = versionSpec.sha256;
+  # };
+
+  buildInputs = [ go-ethereum ];
+
+  # this intentionally reproduces the same directory structure as it
+  # is in the upstream repo, but only copies the files that are
+  # needed.
+  buildCommand = ''
+    mkdir -p $out/lib/systemd/system
+    mkdir -p $out/packaging/nix
+    sed -e 's|ExecStart=.*||' \
```

Comment:
I think we should really use a patch or patches here because with any update anything could change here and it would be ignored.

Reply:
@SuperSandro2000 some of these `sed`s are not static. i.e. they are sometimes referring to `$out`, e.g. in `$out/packaging/4byte.json` (it's a config file that is not supposed to be edited, and thus i put it into the nix store).

IIUC, this makes using patch files impossible. am i right here? shall it remain as it is? (i'm new to all this)

Reply:
@attila-lendvai https://github.com/NixOS/nixpkgs/pull/109647/files is a recent, simple PR that uses dynamic patching. It should provide a decent starting point.

Reply:
@deviant thanks, i tried this, but i make very slow progress, and in an unknown direction to be honest... i assume you suggest this because it's better in some way that i don't see.

is `patches` and `substituteAll` run prior to `buildCommand`?

or does defining a `buildCommand` completely override the application of patches?

i want to only package a few files from $src, hence my `buildCommand`, to only `cp` a few files.

Reply:
@attila-lendvai `stdenv.mkDerivation` has a set of phases it will go through when building a package. First of all there's the `unpackPhase`, which extracts `src` into a directory if it's an archive (here it's not). Then comes the `patchPhase`, where all of the `patches` are applied. `configurePhase` comes next, but it's only relevant if you're building an autotools-style package (and you're not). `buildPhase` is after that, but since you're not actually building anything (just installing it into `$out`), this can be skipped as well. `checkPhase` is where any tests would be run on what was just built in `buildPhase`; again, irrelevant here. `installPhase` is where you actually put files into `$out`. `fixupPhase` is where nixpkgs-specific tweaks are applied (like `patchelf` and wrapper scripts). You can learn more about all of this [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh) and in the nixpkgs manual under the section about phases.

So: in order, unpack -> patch -> configure -> build -> check -> install -> fixup

The suggestion to use patches instead of sed is because the underlying files might change between versions, possibly resulting in your sed invocations no longer doing anything, but you being unaware of this. Say if the `--configdir` option was renamed to `--config-dir`. The sed would simply not find any of these instances, not apply any replacements as a result, but still exit successfully — as nothing about the command actually failed. Patches error out loudly if they no longer apply.

Since patches are static (as you said, they can't refer to `$out`), we need to craft a patch *template* and then substitute in our dynamic values. That's what the `substituteAll` does: it takes a file with a series of placeholders (formatted like `@foo@`, `@bar@`, etc), and fills them in. In the example I gave you, that's what the `inherit` is doing: it's the same as calling `substituteAll { src = ./fix-paths.patch; swaybg = <swaybg from surrounding scope>; }`, which then fills in the `@swaybg@` placeholder with the full path to that package's output.

So: you get a more robust package recipe, while still being able to incorporate dynamic content.

(Note: I have not reviewed whether these are the correct substitutions to make, or whether this package is sane/etc; I just noticed that you were wondering how to do something I had done recently, and thought you'd benefit from seeing it.)

Reply:
@deviant thank you very much for the detailed description! now i know what i should do and why.

i'm still struggling though:

```
  patches = [
    (substituteAll {
      src = ./bee-clef.patch;
      clef = "${go-ethereum}/bin/clef";
      # this fails, but i also need to somehow inject paths that are referring to files in $out 
      #??? inherit out;
    })
  ];
```

this makes the patch apply (and indeed fail loudly when it doesn't apply).

in the patch i do this: 
```
@clef@ ... --rules @out@/packaging/rules.js ...
```

but i get this (notice that $out got replaced to the **patch file**, not the bee-clef package's output):

```
/nix/store/ahqyg0cybhiv15jgh37ydi7yv1md79hw-go-ethereum-1.9.24/bin/clef ... --rules /nix/store/2w3qj5dw172vbjbq66gv2wx0rvkqknyr-bee-clef.patch/packaging/rules.js ...
```

Reply:
(i pushed what i have locally)